source/opt/optimizer.hpp

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SPIRV_TOOLS_OPTIMIZER_HPP_
+#define SPIRV_TOOLS_OPTIMIZER_HPP_
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "libspirv.hpp"
+#include "message.h"
+
+namespace spvtools {
+
+// C++ interface for SPIR-V optimization functionalities. It wraps the context
+// (including target environment and the corresponding SPIR-V grammar) and
+// provides methods for registering optimization passes and optimizing.
+//
+// Instances of this class provides basic thread-safety guarantee.
+class Optimizer {
+ public:
+  // Constructs an instance with the given target |env|, which is used to decode
+  // the binaries to be optimized later.
+  //
+  // The constructed instance will have an empty message consumer, which just
+  // ignores all messages from the library. Use SetMessageConsumer() to supply
+  // one if messages are of concern.
+  explicit Optimizer(spv_target_env env);
+
+  // Disables copy/move constructor/assignment operations.
+  Optimizer(const Optimizer&) = delete;
+  Optimizer(Optimizer&&) = delete;
+  Optimizer& operator=(const Optimizer&) = delete;
+  Optimizer& operator=(Optimizer&&) = delete;
+
+  // Destructs this instance.
+  ~Optimizer();
+
+  // Sets the message consumer to the given |consumer|. The |consumer| will be
+  // invoked once for each message communicated from the library.
+  void SetMessageConsumer(MessageConsumer consumer);
+
+  // Optimizes the given SPIR-V |binary| module.
+  // Returns true on successful optimization, no matter the module is modified
+  // or not. Returns false if errors occur when processing |binary| using any
+  // of the registered passes. All passes registered after the error-reporting
+  // pass will be skipped. And the contents in |binary| may be invalid.
+  //
+  // Use the Register*Pass() methods in the following to register passes first.
+  // Passes will be run in the exact order of registration.
+  bool Optimize(std::vector<uint32_t>* binary) const;
+
+  // Registers a null pass to this optimizer.
+  // A null pass does nothing to the SPIR-V module to be optimized.
+  Optimizer& RegisterNullPass();
```

Comment:
I don't like that every time we add a new optimization we change the ABI to this class.
I'd rather split this into two parts. This Optimizer class accumulates registered transforms/passes via a single method, e.g. 
```
Optimizer& RegisterTransform(std::unique_ptr<Optimizer::Transform>)
```
An Optimizer::Transform is just a forward-declared struct.

Then have free functions that return such objects: 
```
std::unique_ptr<Optimizer::Transform> spvtools::NullTransform();
std::unique_ptr<Optimizer::Transform> spvtools::SetSpecConstantDefaultValueTransform( blah blah blah);
```

(I'm only using "Transform" in case you want to avoid the name "Pass".  But "Pass" would work as well.)